### ISSUE TITLE
Race condition in JSON pretty print during batch testing.

### ISSUE THREAD

**Issue by yan-foto** [drwetter/testssl.sh#1514]
When doing parallel testing with a big number of host names (> 1000) involving multiple time consuming tests (see below), sometimes after the JSON file is actually finalized additional results are appended to the file rendering it invalid!

1. uname -a
   `Darwin hampers.local 18.7.0 Darwin Kernel Version 18.7.0: Thu Jan 23 06:52:12 PST 2020; root:xnu-4903.278.25~1/RELEASE_X86_64 x86_64`

2. testssl version from the banner: testssl.sh -b 2>/dev/null | head -4 | tail -2
  ```
  testssl.sh       3.1dev fromhttps://testssl.sh/dev/
    (64fea03 2020-02-15 15:22:22 -- )
   ```

3. git log | head -1 (if running from git repo)
  `commit 64fea03f665bed7b03bbc217496f63ee93a26b19`

4. openssl version used by testssl.sh: testssl.sh -b 2>/dev/null | awk -F':' '/openssl/ { print $2}'
  `Using "OpenSSL 1.0.2-chacha (1.0.2k-dev)" [~183 ciphers]`

5. steps to reproduce: testssl.sh or docker command line, if possible incl. host
 ```bash
testssl.sh \
  --cipher-per-proto \
  --grease \
  --server-defaults \
  --server-preference \
  --vulnerable \
  --pfs \
  --assume-http \
  --parallel \
  --file $DOMAIN_LIST \
  --jsonfile-pretty $TARGET_DIR
```

6. what exactly was happening, output is needed
In a sample of > 1K domains at least ~10% produce invalid JSON by appending residual test results to the finalized JSON file such as the following (excerpts only):

```json
"cipherTests"       : [

                    ]
          }
          ],
                    "scanTime"  : 633
}
                            {
                                "id"           : "cipher-tls1_xc013",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   xc013   ECDHE-RSA-AES128-SHA              ECDH 256   AES         128      TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_x2f",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x2f     AES128-SHA                        RSA        AES         128      TLS_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "scanProblem",
                                "severity"     : "FATAL",
                                "finding"      : "repeated openssl s_client connect problem, doesn't make sense to continue"
                           }
```

or 

```json
"cipherTests"       : [
                            {
                                "id"           : "cipher-tls1_xc014",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   xc014   ECDHE-RSA-AES256-SHA              ECDH 256   AES         256      TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_x39",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x39     DHE-RSA-AES256-SHA                DH 1024    AES         256      TLS_DHE_RSA_WITH_AES_256_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_x35",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x35     AES256-SHA                        RSA        AES         256      TLS_RSA_WITH_AES_256_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_xc013",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   xc013   ECDHE-RSA-AES128-SHA              ECDH 256   AES         128      TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_x33",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x33     DHE-RSA-AES128-SHA                DH 1024    AES         128      TLS_DHE_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_x2f",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x2f     AES128-SHA                        RSA        AES         128      TLS_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_x05",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x05     RC4-SHA                           RSA        RC4         128      TLS_RSA_WITH_RC4_128_SHA"
                           },{
                                "id"           : "cipher-tls1_x04",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x04     RC4-MD5                           RSA        RC4         128      TLS_RSA_WITH_RC4_128_MD5"
                           },{
                                "id"           : "cipher-tls1_x0a",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1   x0a     DES-CBC3-SHA                      RSA        3DES        168      TLS_RSA_WITH_3DES_EDE_CBC_SHA"
                           }
                    ]
          }
          ],
                    "scanTime"  : "Scan interrupted"
}
,{
                                "id"           : "cipher-tls1_1_xc014",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   xc014   ECDHE-RSA-AES256-SHA              ECDH 256   AES         256      TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_1_x39",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x39     DHE-RSA-AES256-SHA                DH 1024    AES         256      TLS_DHE_RSA_WITH_AES_256_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_1_x35",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x35     AES256-SHA                        RSA        AES         256      TLS_RSA_WITH_AES_256_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_1_xc013",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   xc013   ECDHE-RSA-AES128-SHA              ECDH 256   AES         128      TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_1_x33",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x33     DHE-RSA-AES128-SHA                DH 1024    AES         128      TLS_DHE_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_1_x2f",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x2f     AES128-SHA                        RSA        AES         128      TLS_RSA_WITH_AES_128_CBC_SHA"
                           },{
                                "id"           : "cipher-tls1_1_x05",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x05     RC4-SHA                           RSA        RC4         128      TLS_RSA_WITH_RC4_128_SHA"
                           },{
                                "id"           : "cipher-tls1_1_x04",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x04     RC4-MD5                           RSA        RC4         128      TLS_RSA_WITH_RC4_128_MD5"
                           },{
                                "id"           : "cipher-tls1_1_x0a",
                                "severity"     : "INFO",
                                "finding"      : "TLS 1.1   x0a     DES-CBC3-SHA                      RSA        3DES        168      TLS_RSA_WITH_3DES_EDE_CBC_SHA"
                           },{
                                "id"           : "scanProblem",
                                "severity"     : "FATAL",
                                "finding"      : "repeated openssl s_client connect problem, doesn't make sense to continue"
                           }
```

As you can see, in the former example what should have been inside the `cipherTests` is printed after the JSON file is closed and in the latter case after the test is interrupted some residual results are appended to the file.

---

**Note:** I observed this behavior *after* adding extra tests (i.e., `--grease`, `--server-defaults`, and`--server-preference`, however, I cannot be sure if this is the reason as files are all valid when testing single hosts.

You can take a look at some examples where this happens [here](https://gist.github.com/yan-foto/98cac8d7ee6cbdef34258cf7976b2c22).

**Comment by yan-foto** [comment]
I have been playing around a bit with this problem and I think it should easily be reproduced as follows:

 - Reduce `$MAX_WAIT_TEST` to make sure that the test times out
 - Select a host name with multiple IP addresses (if I'm right, time out is per test and not per IP)
 - Select as many tests as possible (e.g., `--cipher-per-proto --greas
e --protocols --server-defaults --server-preference --vulnerable --pfs`)!

I suppose this happens somewhere around `cleanup()`, but I lack enough experience to pinpoint the exact problem!

**Comment by drwetter** [comment]
@dcooper16 : you have an idea?

I'll be back in a week or a bit more.

**Comment by dcooper16** [comment]
> @dcooper16 : you have an idea?

Not at the moment. I just tried some experiments and was not able to reproduce these results, but I did discover something that _may_ be relevant.

I tried running testssl.sh as
```
bash -x testssl.sh --cipher-per-proto --client-simulation --jsonfile-pretty test.json testssl.sh
```
If I run this on Ubuntu and hit \<Ctrl\>-C while the cipher-per-proto test is being performed, execution immediately jumps to `cleanup()` which closes out the JSON file and then exits.

If I run the above on OS X, hitting \<Ctrl\>-C just results in the program jumping out of the `while` loop in `run_cipher_per_proto()`, but then continuing with the `return` statement at the end of `run_cipher_per_proto()`. So, the client simulations are run and `cleanup()` isn't called until after this.

At the moment, I don't see how that could lead to the results that @yan-foto is seeing, but it suggests that this may be an OS specific problem.

**Comment by yan-foto** [comment]
@dcooper16 that sounds plausible. Does hitting <Ctrl>-C have the same effect as when tests time out?

**Comment by dcooper16** [comment]
I did some more testing and found a way to reproduce the results on both OS X and Ubuntu. I set MAX_PARALLEL to 2 and MAX_WAIT_TEST to 20. I then ran a test as:
```
testssl.sh --cipher-per-proto --client-simulation --jsonfile-pretty jsondir --file domainstotest.txt
```
This results in the types of outputs that @yan-foto has been seeing.

I then modified the line `run_mass_testing_parallel()` that creates child processes from:
```
CHILD_MASS_TESTING=true "$RUN_DIR/$PROG_NAME" "${MASS_TESTING_CMDLINE[@]}" > "$TEMPDIR/term_output_$(printf "%08d" $NR_PARALLEL_TESTS).log" 2>&1 &
```
to
```
CHILD_MASS_TESTING=true bash -x "$RUN_DIR/$PROG_NAME" "${MASS_TESTING_CMDLINE[@]}" > "$TEMPDIR/term_output_$(printf "%08d" $NR_PARALLEL_TESTS).log" 2>&1 &
```
Adding the `bash -x ` caused the child processes to create debugging output.

The debugging output for one of the child processes showed `cleanup()` being executed in parallel with `ciphers_by_strength()` and the program didn't stop running when `cleanup()` finished.

I'll look into this some more later. Perhaps when we call the `kill` command we need to specify a signal on the command line -- one that will give us the expected behavior of the killed process just executing `cleanup()` and then exiting.
